### PR TITLE
Re-implement flapping

### PR DESCRIPTION
**Pull request by Crunsher**
Todo:
* MORE tests
* One of the tests fails, idk why though
* Documentation

### CODE REVIEW

**Comment by dnsmichi** [comment]
A short remark about compatibility - we should keep `flapping_threshold` but immediately deprecate it in 2.8 (no functionality at all). There's a commit sitting in #5682 which adds "deprecated" to mkclass. Wait on that.

**Comment by Crunsher** [comment]
I think we should make an extra pull request for that or put it in this one, it does not have much to do with pki :D

**Comment by dnsmichi** [comment]
Merge to master, rebase this PR against master, profit :)

**Comment by Crunsher** [comment]
It seems like FlappingStart notifications aren't sent.

**Comment by Crunsher** [comment]
enable_flapping is still default off btw

**Comment by dnsmichi** [comment]
It should be left as off, and explicitly enabled if needed imho.

File: lib/icinga/compatutility.cpp [hunk 1]
```diff
@@ -300,12 +300,12 @@ int CompatUtility::GetCheckableIsVolatile(const Checkable::Ptr& checkable)
 
 double CompatUtility::GetCheckableLowFlapThreshold(const Checkable::Ptr& checkable)
```
Comment (gunnarbeutner) [inline lib/icinga/compatutility.cpp]: Can we please remove those CompatUtility::* methods? :)

File: lib/icinga/checkable.ti [hunk 2]
```diff
@@ -70,9 +70,7 @@ abstract class Checkable : CustomVarObject
 		}}}
 	};
 	[config] bool volatile;
-	[config] double flapping_threshold {
```
Comment (gunnarbeutner) [inline lib/icinga/checkable.ti]: We should add a note to the release notes that this config attribute was removed.

File: test/icinga-checkable-flapping.cpp [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	cr->SetScheduleEnd(now);
+	cr->SetExecutionStart(now);
+	cr->SetExecutionEnd(now);
+
+	Utility::IncrementTime(60);
+
+	return cr;
+}
+
+void LogFlapping(const Checkable::Ptr& obj)
```
Comment (gunnarbeutner) [inline test/icinga-checkable-flapping.cpp]: This should probably be static.

File: test/icinga-checkable-flapping.cpp [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+{
+	std::bitset<20> stateChangeBuf = obj->GetFlappingBuffer();
+	int oldestIndex = (obj->GetFlappingBuffer() & 0xFF00000) >> 20;
+
+    std::cout << "Flapping: " << obj->IsFlapping() << "\nHT: " << obj->GetFlappingThresholdHigh() << " LT: " << obj->GetFlappingThresholdLow()
+       << "\nOur value: " << obj->GetFlappingCurrent() << "\nPtr: " << oldestIndex << " Buf: " << stateChangeBuf << '\n';
+}
+
+
+void LogHostStatus(const Host::Ptr &host)
```
Comment (gunnarbeutner) [inline test/icinga-checkable-flapping.cpp]: This should probably also be static.

File: lib/icinga/checkable.ti (hunk 2 above)
Comment (Crunsher) [reply lib/icinga/checkable.ti]: Done with new commit

File: lib/icinga/compatutility.cpp (hunk 1 above)
Comment (Crunsher) [reply lib/icinga/compatutility.cpp]: This is a different topic

**Comment by Crunsher** [comment]
Using the deprecated keyword from mkclass does not work with double, reminder to self to check tomorrow

**Pull request merged**